### PR TITLE
Set more specific batch bundle action in audit event when possible #1817

### PR DESCRIPTION
**Pull request by prb112**
Signed-off-by: Paul Bastide <pbastide@us.ibm.com>

### CODE REVIEW

File: fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java [hunk 1]
```diff
@@ -368,29 +373,55 @@ public static void logBundle(HttpServletRequest request, Bundle requestBundle, B
                 for (Entry bundleEntry : requestBundle.getEntry()) {
                     if (bundleEntry.getRequest() != null && bundleEntry.getRequest().getMethod() != null) {
                         requestMethod = bundleEntry.getRequest().getMethod();
+                        boolean operation =  bundleEntry.getRequest()
+                                                .getUrl().getValue().contains("$");
```
Comment (lmsurpre) [inline fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java:377]: I'm assuming that the URL will always be unencoded here and so its safe to rely on `$` and not `%24`?

File: fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java [hunk 2]
```diff
@@ -368,29 +373,55 @@ public static void logBundle(HttpServletRequest request, Bundle requestBundle, B
...
+                                executeCount++;
+                            } else {
+                                createCount++;
+                            }
                             break;
                         case PUT:
-                            updateCount++;
+                            if (operation) {
+                                executeCount++;
+                            } else {
```
Comment (lmsurpre) [inline fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java:396]: I don't think operations can be invoked via PUT

File: fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java [hunk 3]
```diff
@@ -368,29 +373,55 @@ public static void logBundle(HttpServletRequest request, Bundle requestBundle, B
...
+                                updateCount++;
+                            }
+                            break;
+                        case DELETE:
+                            if (operation) {
+                                executeCount++;
+                            } else {
```
Comment (lmsurpre) [inline fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java:403]: I don't think operations can be invoked via DELETE

File: fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java [hunk 4]
```diff
@@ -368,29 +373,55 @@ public static void logBundle(HttpServletRequest request, Bundle requestBundle, B
...
+                    .resourcesDeleted(deleteCount)
+                    .resourcesExecuted(executeCount)
+                    .build());
             entry.setDescription("FHIR Bundle request");
 
-            // Team discussion results in a recommendation of 'E'
-            // New logic should ensure consistency, all R, all U, all D, all C
-            // when mixed default to E
-            entry.getContext().setAction("E");
+            //
```
Comment (lmsurpre) [inline fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java:423]: either remove or summarize the behavior.  i think better to remove and add any pertinent info to the selectActionForBundle javadoc
```suggestion
```

File: fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java [hunk 5]
```diff
@@ -368,29 +373,55 @@ public static void logBundle(HttpServletRequest request, Bundle requestBundle, B
...
+                                createCount++;
+                            }
                             break;
                         case PUT:
-                            updateCount++;
+                            if (operation) {
+                                executeCount++;
+                            } else {
+                                updateCount++;
+                            }
```
Comment (prb112) [inline fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java:398]:
```suggestion
                            updateCount++;
```

File: fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java (hunk 2 above)
Comment (prb112) [reply fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java:396]: I double checked the code, you are right.

File: fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java (hunk 3 above)
Comment (prb112) [reply fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java:403]: Yes, they can - I added it for bulkdata-status.

File: fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java (hunk 1 above)
Comment (prb112) [inline fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java:377]:
```suggestion
                        boolean operation =  bundleEntry.getRequest().getUrl().getValue().contains("$")
                                                || bundleEntry.getRequest().getUrl().getValue().contains("/%24");
```

File: fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java (hunk 4 above)
Comment (prb112) [reply fhir-server/src/main/java/com/ibm/fhir/server/util/RestAuditLogger.java:423]: I ended up creating the helper method instead.

**Review by lmsurpre — APPROVED** [review]
LGTM

**Pull request merged**